### ISSUE TITLE
Support system user in Storage

### ISSUE THREAD

**Issue by SandGrainOne** [Altinn/altinn-storage#501]
### Description

This is a list of tasks and changes that needs to be done before we can say that Storage support system users.

### Additional Information

This issue is based on the analysis performed in #471.

Usecase: A system user (representing a user - i.e. sending a party-id) instantiates an instance and populates it with data. E.g. The MVA-report.

#### Authorizaion
- [x] Update to a version of the [Altinn.Common.PEP](https://www.nuget.org/packages/Altinn.Common.PEP#versions-body-tab) package with support for SystemUser authorization (version later than v4.0.0)
- [x] Update all custom Authorization logic to include system user as AccessSubject in Authorization requests.
    The assumption is that we're mostly using the PEP package and the containing DecisionHelper, but there might be exceptions.

#### Metadata updates:
- [x] Search for places where the Intance.CreatedBy property is set. Update the logic to use the organization number of the system owner when caller is a system.
- [x] Search for places where the Intance.LastChangedBy property is set. Update the logic to use the organization number of the system owner when caller is a system.
- [x] Search for places where the DataElement.CreatedBy property is set. Update the logic to use the organization number of the system owner when caller is a system.
- [x] Search for places where the DataElement.LastChangedBy property is set. Update the logic to use the organization number of the system owner when caller is a system.
- [x] Expand the PlatformUser class with properties for: SystemUserId (guid), SystemUserOwner (string) and SystemUserName.
- [x] Search for all usage (assignments) of the PlatformUser class and populate the new properties when data is available.
- [x] Update the logic setting the ProcessHistoryItem.PerformedBy property to support an InstanceEvent created with a system user.
- [x] Update IdentityTelemetryFilter to handle system users

#### Authentication
Not decided yet (optional future improvement):
- [ ] https://github.com/Altinn/altinn-storage/issues/509

### Preparation
Testing this will require multiple steps. 
1. Find a person with the ability to represent an organization. Preferably the "CEO" (daglig leder). Make node of key data we might need. Like names and ids. There are a few Tenor users we've documented on confluence under Guides.
2. We would then need to create a system user. I have not looked into how this could be done yet. The system user would need to be associated with the organization we've identified previously. It doesn't matter what type of system we use. Keep the data like system id and update the test data table on confluence.
3. Find a working app and instantiate an instance or two with the organization as instance owner. (Or find existing instances).
4. Log in with the organization CEO and delegate at a minimum read access to the system user. I have not looked into how this is done at the moment.
5. Generate a token for the system user using the token generator. There is a new endpoint to do this part. An example:
`api/GetSystemUserToken?env={{Environment}}&scopes=altinn:instances.read&systemUserId=3d7ff126-890c-4e30-953b-4b736c6591ec&systemUserOrg=312508729`
6. Run GET requests directly against Storage endpoints using the token as a Bearer token in the Authorization header.

### Things to check
1. First of all the GET requests should succeed.
2. Logging of system user id together with request telemetry. Check in Application Insights.

Testing of mutating operations like POST and PUT is even more work and I suggest we postpone that for when we have a working app to test through.

```[tasklist]
### Blockers
- [ ] https://github.com/Altinn/altinn-authorization/issues/578
- [ ] https://github.com/Altinn/altinn-authorization/pull/1052
```

```[tasklist]
### Tasks
- [x] Get initial session from Rune for building competency
- [ ] https://github.com/Altinn/altinn-storage/issues/511
- [ ] https://github.com/Altinn/altinn-storage/issues/512
- [ ] https://github.com/Altinn/altinn-storage/issues/513
- [ ] https://github.com/Altinn/altinn-storage/issues/510
```

### Acceptance criteria:
- [ ] Component accepts and responds to calls from system users
  - [ ] If feature-flag "off", the component gracefully denies calls from system users
- [ ] Component behaves as before for calls from other users

**Comment by olebhansen** [comment]
FYI the competency-session with Rune is scheduled. Ticking off the box and unassigning myself.

**Comment by Ahmed-Ghanam** [comment]
**Monday, 03 February 2025**

A system user has been created under the organization **312508729**, which holds the party identifier **51567369**. This system user has full access to the application **storage end-to-end** and is identified by **`a3a80036-2c2c-4c7b-91ca-6132493b4a0f`**. This system user was utilized to generate a bearer token, which was subsequently used to call all GET endpoints.

### **Accepted Endpoints**

#### **Applications**
- **GET** `/applications`
- **GET** `/applications/{org}/{app}`

#### **Data**
- **GET** `/instances/{instanceOwnerPartyId}/{instanceGuid}/data/{dataGuid}`
- **GET** `/instances/{instanceOwnerPartyId}/{instanceGuid}/events/{eventGuid}`
- **GET** `/instances/{instanceOwnerPartyId}/{instanceGuid}/dataelements`

#### **Instance Events**
- **GET** `/instances/{instanceOwnerPartyId}/{instanceGuid}/events`
- **GET** `/instances/{instanceOwnerPartyId}/{instanceGuid}/events/{eventGuid}`

#### **Instances**
- **GET** `/instances/`
- **GET** `/instances/{instanceOwnerPartyId}/{instanceGuid}`

#### **MessageBox Instances**
- **GET** `/sbl/instances/{instanceOwnerPartyId}/{instanceGuid}`
  - **Result:** *"Access denied. Operation requires a subscription associated with the SblAccess product."*
- **GET** `/instances/{instanceOwnerPartyId}/{instanceGuid}/events`
  - **Result:** *"Access denied. Operation requires a subscription associated with the SblAccess product."*

#### **Texts**
- **GET** `/applications/{org}/{app}/texts/{language}`

**Comment by Ahmed-Ghanam** [comment]
**Tuesday, 04 February 2025**
I invoked the `GET /instances/` endpoint using a system user with a valid bearer token. Upon reviewing the request telemetry, I observed that the system user identifier and the system user organization identifier were recorded as part of the telemetry data.

**Comment by Ahmed-Ghanam** [comment]
This issue has been tested in the AT22 test environment and successfully validated. It is now ready for closure. Excellent work.
